doc/source/cluster/kubernetes/examples/distributed-checkpointing-with-gcsfuse.md

```diff
@@ -0,0 +1,264 @@
+```
+export PROJECT_ID=<your project id>
+gcloud container clusters create kuberay-with-gcsfuse \
+    --addons GcsFuseCsiDriver \
+    --cluster-version=1.29.3-gke.1093000 \
+    --location=us-east4-c \
+    --machine-type=g2-standard-8 \
+    --num-nodes=4 \
+    --accelerator type=nvidia-l4,count=1,gpu-driver-version=latest \
+    --workload-pool=${PROJECT_ID}.svc.id.goog
+```
+
+## Install the KubeRay operator
+
+Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.
+The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.
+
+## Configuring the GCS Bucket
+
+Create a GCS bucket that will be used as the remote filesystem.
+```
+gcloud storage buckets create gs://<your bucket name>
+```
+
+Create a Kubernetes ServiceAccount that will be used to grant the RayCluster access to mount the GCS bucket:
+```
+kubectl create serviceaccount pytorch-distributed-training
+```
+
+Bind the `roles/storage.objectUser` role to the Kubernetes service account and bucket IAM policy:
+```
+PROJECT_ID=<your project ID>
+PROJECT_NUMBER=<your project number>
+BUCKET=<your GCS BUCKET>
+gcloud storage buckets add-iam-policy-binding gs://${BUCKET} --member "principal://iam.googleapis.com/projects/${PROJECT_NUMBER}/locations/global/workloadIdentityPools/${PROJECT_ID}.svc.id.goog/subject/ns/default/sa/pytorch-distributed-training"  --role "roles/storage.objectUser"
+```
+
+See [Access Cloud Storage buckets with the Cloud Storage FUSE CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) for more details.
+
+## Deploy the RayJob
+
+Download the RayJob that executes all the steps documented in [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html). The [source code](https://github.com/ray-project/kuberay/tree/master/ray-operator/config/samples/pytorch-resnet-image-classifier) is also in the KubeRay repository.
+
+```
+curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml
+```
+
+Modify the RayJob by replacing all instances of the `GCS_BUCKET` placeholder with the Google Cloud Storage bucket you created earlier. Alternatively you can use `sed`:
+```
+sed -i "s/GCS_BUCKET/<your-bucket>/g" ray-job.pytorch-image-classifier.yaml
+```
+
+Deploy the RayJob
+```
+kubectl create -f ray-job.pytorch-image-classifier.yaml
+```
+
+The deployed RayJob includes the following configuration to enable distributed checkpointing to a shared filesystem:
```

Comment:
Would you mind adding more details about the following configurations in the YAML? Currently, users can't understand how the storage looks.

```yaml
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
...
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

Reply:
I don't want to go into too much detail about all the possible annotations and mount options. I added this with some links to GCP docs:
```
The GCSFuse volume is configured with annotations which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.
```
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
```

Mount options are also specified when defining the GCSFuse container volume:
```
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

See [Mount options](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#mount-options) to learn more about mount options.
```